### DISCUSSION TITLE
Trying to patch unison

### DISCUSSION THREAD

**Discussion by brennongs** [Homebrew/discussions#1174] [Everyday usage] [answered]
Yesterday my team's Docker scripts all broke! We're using `docker-sync` with the `unison` strategy. After much digging, I found it's because there's an outdated patch in the `unison` formula. I'm trying to update the patch, and having trouble. The fix is in a release candidate, and the `livecheck` is yelling at me, but I dunno if I should monkey with that or not. I have tried a lot of different ways to do the patch, but whenever I build from source, I always get one of two errors. either "Garbage was found in the patch input" or a `MissingApplyError`. I can't find any documentation about either of them. Please help! I would love to be able to push this fix and make my measly contribution to this awesome tool but I'm feeling stuck :(

**Comment by carlocab** [comment, marked as answer]
In what way is the patch outdated? It looks like the references are all correct. Does it no longer apply? Are you installing from the bottle, or building from source? If you're installing from the bottle then whether the patch is outdated or not shouldn't be affecting you.

**Reply by brennongs** [reply]
I may be misunderstanding the patch comments, but the comment says it is valid until merge 480 or merge 481 has been merged and released in unison. 481 was merged and is in a release candidate. I have successfully installed the code from 4_rc2 twice (although both times I know it was not the brew way, which is why I'm still struggling), and both times our docker setup was fixed. To be honest, I'm not entirely sure why it is broken in the first place, but it's the only thing I've found that works.

I'm building from source. Is `installing from the bottle` the same as just doing `brew install unison`? because when we just `brew install unison`, our unison is broken.

**Reply by SMillerDev** [reply]
> 481 was merged and is in a release candidate

Homebrew/core requires releases to be marked stable. So brew audit will always complain when you feed a release candidate

**Reply by brennongs** [reply]
Hmm that makes sense. Is there a way my team and I can temporarily install the release candidate or roll back to the last version until the official release?

**Reply by carlocab** [reply]
It's still in a release candidate, though, which isn't the version that Homebrew ships. The patch still applies to the version that Homebrew ships (2.51.3).

> I'm building from source. Is `installing from the bottle` the same as just doing `brew install unison`? because when we just `brew install unison`, our unison is broken.

Yes, `brew install unison` should be installing from a bottle (if you're on Mojave, Catalina, or Big Sur). Though now I'm not sure why you say you're building from source, if it's just `brew install unison`.

What exactly does "broken" mean here? Sorry, but that's not really much to go on.

**Reply by carlocab** [reply]
You can try `brew install --HEAD unison`, which will install the version at the tip of the main/master branch.

**Reply by brennongs** [reply]
Sorry, you're right I'm not giving enough context.

We all did `brew install unison` and everything was working. On monday my teammate did `brew update`, and found that unison was timing out when trying to sync (unfortunately I still don't understand enough about how unison works under the hood to really understand more than that). Yesterday I did `brew update` and started getting the same error. In my attempt to fix it, I changed the formula locally to install the unison from 4_rc2, commented out the `livecheck` and hardcoded the version to suppress the failing test error (obviously not the right way to go about it), and when I did `brew uninstall unison --force` and `brew install --build-from-source unison`, our docker setup was back to working the way we expected. So now i've been trying to do the patch in the correct way but I can't find documentation on how to fix the code to not error at me.

I will try to install from head and see if that resolves my issue.

**Reply by SMillerDev** [reply]
It would be really appreciated if you can try and make a PR for a better test for `unison` that actually tries some of the functionality. That will stop us from shipping you a broken version next time 😄

**Reply by carlocab** [reply]
Hm, yes, it's possible the patch introduced a bug that fixed before the release candidate. Sorry about that. Locally, you can just install the release candidate by changing the `url` and `sha256` (and you can just delete the `patch` block entirely). 

Otherwise, you can try `brew install --HEAD` to use the latest development version, but you'll need to uninstall it first to switch back to the stable version when 2.51.4 is shipped in homebrew/core.

**Reply by brennongs** [reply]
I would totally make a PR for the test if I knew where to start, but I have only used unison at a high level for a single thing and I think I would end up just spinning my wheels endlessly 😅. Can I open an issue in some way that reveals that there's an issue so someone who knows more about unison can know there's help needed for that?

Luckily, installing with `--HEAD` worked. Do we just need to keep our eye on the formula to know when version 4 is released? Also some of us have a launchctl job that automatically updates brew for us, is brew update going to write over the head version?

**Reply by carlocab** [reply]
> I would totally make a PR for the test if I knew where to start, but I have only used unison at a high level for a single thing and I think I would end up just spinning my wheels endlessly 😅

That's enough to make you an expert on `unison` compared to us!

> Luckily, installing with `--HEAD` worked. Do we just need to keep our eye on the formula to know when version 4 is released? Also some of us have a launchctl job that automatically updates brew for us, is brew update going to write over the head version?

`brew {update,upgrade}` does not touch any `HEAD` installs, so your `launchctl` jobs should be fine. Any updates to your `unison` install will need to be done manually. So, yes, you'll need to keep an eye on the formula to update to 2.51.4 with `brew uninstall unison && brew install unison`.

You might also want to make sure everyone who uses `unison` installs it at roughly the same time to make sure you're building from the same commit on the master branch.

**Reply by brennongs** [reply]
Ok awesome, thank you for the information.

As for the PR, I will see what I can do. I'm still a little afraid of open source 😂
